File: 2020/Day10/Haskell/AdapterArray-Part2.hs
import System.IO
import Control.Monad
import Data.List(sort, nub)

printArray:: (Show a, Eq a) => [a] -> String
printArray (x:xs)
    | null xs = show x
    | otherwise = show x ++ "\n" ++ printArray xs

countAdapters:: [Int] -> [Int]
countAdapters (x:xs)
    | null xs = [1]
    | otherwise = sum (zipWith const countTail (takeWhile (<= x+3) xs)) : countTail
    where countTail = countAdapters xs  

main:: IO()
main = do
    handle <- openFile "../input.txt" ReadMode
    contents <- hGetContents handle
    let rows = lines contents
    let ad = 0 : sort (map (read::String->Int) rows)
    let adapters = ad ++ [maximum ad + 3]
    print (head $ countAdapters adapters)
    hClose handle

File: 2020/Day10/Python/AdapterArray-Part2.py
import sys
import collections
import itertools
import numpy as np

adapters = [0]
for line in open(f"{sys.path[0]}/../input.txt").read().splitlines():
    adapters.append(int(line))
adapters.sort()
adapters.append(max(adapters)+3)

seen = {}
def countAdapters(ad):
    if ad == adapters[-1]:
        return 1
    elif ad in seen:
        return seen[ad]
    total = 0
    for j in range(adapters.index(ad)+1, len(adapters)):
        if (adapters[j] - ad) <= 3:
            total += countAdapters(adapters[j])
    seen[ad] = total
    return total
    
print(countAdapters(0))